### PR TITLE
Crop during training and padding during inference

### PR DESCRIPTION
**Pull request by apyskir**
Also included changes from @taraspiotr 's PR #85 and removed annoying _pipeline sufffix in main.py

### CODE REVIEW

File: src/augmentation.py [hunk 1]
```diff
@@ -166,4 +166,166 @@ def _perspective_transform_augment_images(self, images, random_state, parents, h
...
+            image_cropped = image[crop_top:crop_bottom, crop_left:crop_right]
+        else:
+            image_cropped = image[crop_top:crop_bottom, crop_left:crop_right, :]
+        return image_cropped
+
+    def get_parameters(self):
+        return []
+
+
+class PadToFitInNetwork(iaa.Augmenter):
```
Comment (jakubczakon) [inline src/augmentation.py]: @apyskir not sure about this name. How about we call it InferencePad or DivisibleSizePad?

File: src/augmentation.py [hunk 2]
```diff
@@ -166,4 +166,166 @@ def _perspective_transform_augment_images(self, images, random_state, parents, h
...
+            result.append(image_padded)
+        return result
+
+    def _pad_image(self, image):
+        height = image.shape[0]
+        width = image.shape[1]
+
+        pad_vertical = self._get_pad(height)
+        pad_horizontal = self._get_pad(width)
+        pad_sequence = self._get_pad_sequence(pad_vertical, pad_horizontal)
```
Comment (jakubczakon) [inline src/augmentation.py]: @apyskir I would hide `self._get_pad(height)` and `self._get_pad(width)` in the `self._get_pad_sequence` since we are not using it outside of that func

File: src/callbacks.py [hunk 3]
```diff
@@ -160,9 +161,15 @@ def _transform(self):
                 break
         self.model.train()
         average_losses = sum(partial_batch_losses) / steps
-        outputs = {'{}_prediction'.format(name): np.vstack(outputs_) for name, outputs_ in outputs.items()}
+        try:
```
Comment (jakubczakon) [inline src/callbacks.py]: @apyskir I don' love using try catch here but handling it with lists here is probably not necessary.

File: src/callbacks.py [hunk 4]
```diff
@@ -207,10 +214,16 @@ def on_epoch_end(self, *args, **kwargs):
         self.epoch_id += 1
 
 
-def postprocessing__pipeline_simplified(cache_dirpath):
+def postprocessing__pipeline_simplified(cache_dirpath, loader_mode):
+    if loader_mode == 'crop_and_pad':
+        resize_function = post.crop_image
```
Comment (jakubczakon) [inline src/callbacks.py]: @apyskir `resize_function` doesn't seem correct here. format_function or size_adjustment_function?

File: src/augmentation.py (hunk 1 above)
Comment (apyskir) [reply src/augmentation.py]: @jakubczakon OK let's keep `InferencePad`

File: src/loaders.py [hunk 5]
```diff
@@ -31,69 +32,33 @@ def __init__(self, X, y, train_mode,
...
         self.image_augment_with_target = image_augment_with_target
 
-    def load_image(self, img_filepath):
-        image = Image.open(img_filepath, 'r')
-        return image.convert('RGB')
+        self.image_source = image_source
 
     def __len__(self):
-        return self.X.shape[0]
+        if self.image_source == 'memory':
```
Comment (jakubczakon) [inline src/loaders.py]: @apyskir of course! good job on cleaning this puppy up

File: src/augmentation.py (hunk 2 above)
Comment (apyskir) [reply src/augmentation.py]: @jakubczakon OK

File: src/callbacks.py (hunk 4 above)
Comment (apyskir) [reply src/callbacks.py]: @jakubczakon Let's keep `size_adjustment_function`

File: src/loaders.py [hunk 6]
```diff
@@ -31,69 +32,33 @@ def __init__(self, X, y, train_mode,
...
-    def __getitem__(self, index):
-        Xi = self.X[0][index]
+        Xi = load_func(self.X, index)
 
         if self.y is not None:
-            Mi = self.y[0][index]
+            Mi = load_func(self.y, index)
 
-            if self.train_mode and self.image_augment_with_target is not None:
+            if self.image_augment_with_target is not None or self.image_augment is not None:
```
Comment (jakubczakon) [inline src/loaders.py]: @apyskir this looks weird. We are checking condition 1 and 2 and inside of that we are checking them again one by one ?

File: src/loaders.py [hunk 7]
```diff
@@ -149,17 +124,19 @@ def get_datagen(self, X, y, train_mode, loader_params):
...
                                    mask_transform=self.mask_transform,
-                                   image_transform=self.image_transform)
+                                   image_transform=self.image_transform,
+                                   image_source=self.dataset_params.image_source)
         else:
             dataset = self.dataset(X, y,
                                    train_mode=False,
-                                   image_augment=None,
-                                   image_augment_with_target=None,
+                                   image_augment=self.image_augment_inference,
```
Comment (jakubczakon) [inline src/loaders.py]: @apyskir mhm exactly how I intended to use it. It seems to me that this option lets one do pretty much any loading acrobations (tta, random aug, prep). Would you agree?

File: src/models.py [hunk 8]
```diff
@@ -83,10 +84,44 @@ def fit(self, datagen, validation_datagen=None, meta_valid=None):
...
+            if len(self.output_names) == 1:
+                outputs.setdefault(self.output_names[0], []).append(outputs_batch.data.cpu().numpy())
+            else:
+                for name, output in zip(self.output_names, outputs_batch):
+                    output_ = output.data.cpu().numpy()
+                    outputs.setdefault(name, []).append(output_)
+            if batch_id == steps:
+                break
+        self.model.train()
+        try:
```
Comment (jakubczakon) [inline src/models.py]: @apyskir I don't like the use of try catch here. I think we should have the private `self._trasnsform()` method return a list and if we want to we can vstack it in the public `self.transform` method. What do you think @apyskir @taraspiotr ?

File: src/pipelines.py [hunk 9]
```diff
@@ -183,39 +191,39 @@ def preprocessing_inference(config, model_name='unet'):
...
+                           'y': ([(reader_inference.name, 'y')], squeeze_inputs_if_needed),
+                           'train_mode': ([('specs', 'train_mode')]),
+                           },
+                  cache_dirpath=config.env.cache_dirpath)
     return loader
 
 
 def mask_postprocessing(model, config, save_output=False):
+    if config.execution.loader_mode == 'crop_and_pad':
+        resize_function = crop_image
```
Comment (jakubczakon) [inline src/pipelines.py]: @apyskir again I would change this naming. Can we somehow reuse things because there seems to be some repetition (feels like it at least)

File: src/postprocessing.py [hunk 10]
```diff
@@ -220,3 +237,11 @@ def drop_small(img, min_size):
 def label(mask):
     labeled, nr_true = ndi.label(mask)
     return labeled
+
+
+def _get_crop_sequence(crop_vertical, crop_horizontal):
```
Comment (jakubczakon) [inline src/postprocessing.py]: @apyskir this function was also used in the `augmentations.py` just have it as utils function and use in both places

File: src/steppy/preprocessing/misc.py [hunk 11]
```diff
@@ -34,10 +34,9 @@ def save(self, filepath):
 
 
 class ImageReader(BaseTransformer):
-    def __init__(self, x_columns, y_columns, target_shape):
+    def __init__(self, x_columns, y_columns):
```
Comment (jakubczakon) [inline src/steppy/preprocessing/misc.py]: @apyskir I agree think it could be usefull to have the reshape option in the reader especially since there are some libs that facilitate read+resize that are really fast (fastjpeg or something if I remember correctly). That being said target_shape seems to be obsolete in this implementation.

File: src/utils.py [hunk 12]
```diff
@@ -357,6 +370,19 @@ def get_arg_length(*args, **kwargs):
     return StaticApplyTransformer()
 
 
+def get_seed():
```
Comment (jakubczakon) [inline src/utils.py]: @apyskir this function shows where imgaug gets tricky right?

File: src/loaders.py (hunk 6 above)
Comment (apyskir) [reply src/loaders.py]: @jakubczakon Yes, if any of the augmentations is not None, we want to transform to PIL, and then we want to perform the augmentations if they are not None. I can propose, that I will write:
```
if self.augment_image is None:
   self.augment_image = iaa.Noop()
```
This way we can remove checking the conditions inside __get_item__

File: src/postprocessing.py (hunk 10 above)
Comment (apyskir) [reply src/postprocessing.py]: @jakubczakon OK, so either I put it in `utils.py`, or put `self.get_pad` inside this function in `augmentations.py`. I prefer the former.

File: src/steppy/preprocessing/misc.py (hunk 11 above)
Comment (apyskir) [reply src/steppy/preprocessing/misc.py]: @jakubczakon I removed it, because reshape was done twice, actually: once in the reader, and once in `self.image_transform` in loader.

File: src/pipelines.py (hunk 9 above)
Comment (apyskir) [reply src/pipelines.py]: @jakubczakon Repetition, you mean the same thing in mask_postprocessing, postprocessing_masks and postprocessing_boarders?

File: src/callbacks.py (hunk 3 above)
Comment (apyskir) [reply src/callbacks.py]: @jakubczakon I think that we can drop this try..except here, because batch_size=1 is required, so we can leave it as list.

File: src/loaders.py (hunk 6 above)
Comment (jakubczakon) [reply src/loaders.py]: @apyskir Oh I don't mind checking them in __get_item__ i just think that something unnatural is happening here. For example say we want to just augment_image why do we produce Xi, Mi  and not just Xi ?

File: src/models.py (hunk 8 above)
Comment (jakubczakon) [reply src/models.py]: @apyskir I agree list of image predictions sounds best

File: src/pipelines.py (hunk 9 above)
Comment (jakubczakon) [reply src/pipelines.py]: @apyskir Yeah it seems so. Maybe some `suffix` that changes the name would suffice (no pun intended)

File: src/postprocessing.py (hunk 10 above)
Comment (jakubczakon) [reply src/postprocessing.py]: @apyskir me too. You could also have cake and eat it by creating a private method there that calls _get_pad first and then calls get_crop_sequence function

File: src/steppy/preprocessing/misc.py (hunk 11 above)
Comment (jakubczakon) [reply src/steppy/preprocessing/misc.py]: @apyskir ok

File: src/loaders.py (hunk 6 above)
Comment (apyskir) [reply src/loaders.py]: @jakubczakon Because if you want to augment_image_with target separately, then you have to convert Xi to PIL twice. I think it's the best option to use my proposal and then always convert Xi, Mi to PIL, apply augmentations, and then back to numpy. An alternative is:
```if self.augment_image_with_target is not None:
  Xi, Mi = to_pil(Xi, Mi)
  Xi, Mi = self.augment_image_with_target
  Xi, Mi = from_pil(Xi, Mi)
if self.augment_image is not None:
  Xi = to_pil(Xi)
  Xi = self.augment_image
  Xi = from_pil(Xi)```
Am I right? Or I can check all combinations of self.augment_image and self.augment_image_with_target being `None` and `not None` but I don't think it's necessary.

File: src/callbacks.py (hunk 3 above)
Comment (apyskir) [reply src/callbacks.py]: I was mistaken. `batch_size=1` is required only if crop_and_pad. Otherwise we can have `batch_size>1`.

File: src/postprocessing.py (hunk 10 above)
Comment (apyskir) [reply src/postprocessing.py]: @jakubczakon I like cakes. Let's do it.

File: src/loaders.py (hunk 6 above)
Comment (jakubczakon) [reply src/loaders.py]: @apyskir I like this more, because it is clear what is happening

File: src/pipelines.py (hunk 9 above)
Comment (apyskir) [reply src/pipelines.py]: @jakubczakon I'm afraid we just have 3 different postprocessing pipelines for masks, masks (not the same masks) and borders, so we have to leave as it is
